src/staging/Tabs/ContextAndChildren/Tabs.tsx

```diff
@@ -0,0 +1,100 @@
+import { Children, FC, isValidElement, ReactElement, ReactNode } from "react";
+import { useControlled } from "../shared/useControlled";
+import { TabProvider, useTabs } from "../shared/TabContext";
+
+interface TabsProps {
+  children: ReactNode;
+  activeTab?: string;
+  onTabChange?: (newTab: string) => string;
+}
+
+export const Tabs: FC<TabsProps> = ({ children, activeTab: activeTabProp, onTabChange: onTabChangeProp }) => {
+  const tabInfo = getTabInfo(children);
+
+  const [activeTab, onChange] = useControlled({ value: activeTabProp, initialValue: "" });
+
+  const onTabChange = (newTab: string) => {
+    onChange?.(newTab);
+    onTabChange?.(newTab);
+  };
+
+  return (
+    <TabProvider activeTab={activeTab} onTabChange={onTabChange}>
+      <TabList tabInfo={tabInfo} />
+      {children}
+    </TabProvider>
+  );
+};
+
+const getTabInfo = (children: ReactNode): { value: string; label: ReactNode }[] => {
+  return (
+    Children.toArray(children)
+      // We are going to be aggregating props
+      // all invalid elements need to go
+      .filter(isValidElement)
```

Comment:
Cant we just do this `isValidElement<TabItemProps>` by passing the tab item props type as a generic instead of the casting below?

Reply:
If you didn't need the specific values below then you could. We haven't validated if the child is a `TabItem` or not yet. `isValidElement` only checks that the child was created with `createElement` not that they have the shape or not yet. Using the type argument here would be lying to ourselves in this case.